### ISSUE TITLE
Feature Request: Enable Option to Receive stdin and Output stdout

### ISSUE THREAD

**Issue by rogerluan** [realm/SwiftLint#3636]
### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### New **feature** request

1. Why should this **feature** be added? Share links to existing discussion about what
   the community thinks about this.

I use SwiftLint alongside [SwiftFormat](https://github.com/nicklockwood/SwiftFormat), because although there're rules common between the two, each of these tools contain different rules. I've been using both SwiftLint + SwiftFormat on my team's `precommit` git hook, where I created a custom script that basically "lints files in staging only". However, although we've been using this script for (probably) over 3 years now, it has some flaws. So I found that SwiftFormat added a section that instructs users how to install a precommit hook pretty well: https://github.com/nicklockwood/SwiftFormat#git-pre-commit-hook

It uses https://github.com/hallettj/git-format-staged - which I found quite of an amazing strategy, take some time to read its README :). I believe it's the state of the art and most accurate way to perform such task of formatting staged files.

However, here comes the problem: SwiftLint doesn't work with `git-format-staged` out of the box because (AFAICT) there's no way to make it print the output to stdout, unlike SwiftFormat. In practical terms:

### SwiftFormat stdin option prints result to stdout

```sh
$ echo "if [].count == 0 {\n\n\n\n}" | Pods/SwiftFormat/CommandLineTool/swiftformat stdin
Running SwiftFormat...
if [].count == 0 {}
Swiftformat completed successfully.
```

### SwiftLint stdin option just analyzes the input stream and prints its analysis to stdout

```sh
$ echo "if [].count == 0 {\n\n\n\n\n\n}" | Pods/SwiftLint/swiftlint lint --no-cache --use-stdin --enable-all-rules
<nopath>:1:7: error: Empty Count Violation: Prefer checking `isEmpty` over comparing `count` to zero. (empty_count)
<nopath>:6:1: warning: Vertical Whitespace Violation: Limit vertical whitespace to a single empty line. Currently 5. (vertical_whitespace)
<nopath>:2:1: warning: Vertical Whitespace before Closing Braces Violation: Don't include vertical whitespace (empty line) before closing braces. (vertical_whitespace_closing_braces)
<nopath>:2:1: warning: Vertical Whitespace after Opening Braces Violation: Don't include vertical whitespace (empty line) after opening braces. (vertical_whitespace_opening_braces)
Done linting! Found 4 violations, 1 serious in 1 file.
```

Thus, this feature request consists of enabling us to request SwiftLint to print the output of a stdin, to stdout (same as autocorrect, but printing to stdout instead of in-place at the original file).

## Proposal

```sh
$ echo "if [].count == 0 {\n\n\n\n\n\n}" | Pods/SwiftLint/swiftlint lint --no-cache --use-stdin --use-stdout --enable-all-rules
if [].isEmpty {}
```

Thanks for reading! Looking forward to hearing your thoughts on this feature and its feasibility, as well as if there's any shortcut we can use to make this happen with zero code changes that I may not be aware of 🙏

**Comment by rogerluan** [comment]
Adding some more additional information I just realized:


Taken from https://github.com/nicklockwood/SwiftFormat:


>When using stdin, SwiftFormat does not have access to the file path of the input, so features that rely on the file location (such as inserting the creation date into header comments, or detecting .swiftformat configuration files in the file path) will not work. To solve this, you can provide the file path using the --stdinpath argument:
>
>```sh
>$ cat /path/to/file.swift | swiftformat stdin --stdinpath /path/to/file.swift
>```

This would make sense to apply to SwiftLint as well, if we were to add this new stdin feature to it, because there are some rules that do depend on the path of the file, e.g. https://realm.github.io/SwiftLint/file_name.html

**Comment by iliaskarim** [comment]
I agree. swfitlint should use stdout.

**Comment by tsapeta** [comment]
@jpsim @marcelofabri are there any plans to implement this? My use case is a bit different, but what I'm doing now is to apply autocorrection, read the file and then revert it. It works, but I think the developer experience could be improved here. It'd love to see it implemented some day 😉

**Comment by SimplyDanny** [comment]
With #4132 stdout is now always used if input is given in stdin.

**Comment by rogerluan** [comment]
<img width="20" alt="amaze" src="https://user-images.githubusercontent.com/8419048/106630923-05e11b00-655b-11eb-858a-bbbc3096291e.gif"> amazing!! Thank you so much! I'll test this as soon as I can 😁

**Comment by rogerluan** [comment]
#4132 made SwiftLint's stdin/stdout kinda work, but it doesn't fully work with https://github.com/hallettj/git-format-staged yet 😞 

I don't have enough time to debug this right now, but this works for SwiftFormat:

```sh
SWIFTFORMAT=Pods/SwiftFormat/CommandLineTool/swiftformat
git_stage_formatter --formatter \"$SWIFTFORMAT stdin --stdinpath '{}'\" \"*.swift\"
```

But for SwiftLint, this is not working:

```sh
SWIFTLINT=Pods/SwiftLint/swiftlint
git_stage_formatter --formatter \"$SWIFTLINT lint --fix --use-stdin '{}'\" \"*.swift\"
```

I think the key difference here is the lack of a "filepath" parameter in SwiftLint, something that is present in SwiftFormat and other linters (e.g. prettier, from javascript), which is required to perform precommit linting of partially staged files 😬 

Not sure if this is something the SwiftLint community members have interest in pursuing, but it's definitely game changer to be able to automatically format *_partially staged_* files (instead of simply "entire files")

**Comment by SimplyDanny** [comment]
I think you can just leave out the `'{}'` after the `--use-stdin`. SwiftLint doesn't need a file path. It always expects Swift code independent of the file extension.

**Comment by rogerluan** [comment]
Good call, but unfortunately that didn't fix it. The problem happens when committing a partially staged file (which is what [git-format-staged](https://github.com/hallettj/git-format-staged) is made for). The behavior when using it with SwiftLint is that the commit turns out fine, but the local code is left with the old changes. Let me give an example:

I have this file:

```swift
enum MyTest {
    static func doSomething() {
        print("This is just a test")
    }
}
```

I edit it so it looks like this:

```swift
enum MyTest {
    static func doSomething() {
        print("This is just a test")
    }
}

enum MyTest2 {



    static func doSomething()
    {


        print("This is just a test")



    }


}

enum MyTest3 {
    static func doSomething() {
        // Something irrelevant that I won't be committing.
    }
}
```

Then I stage and commit this bit only:

```swift
enum MyTest2 {



    static func doSomething()
    {


        print("This is just a test")



    }


}
```

The commit turns out fine:

<img width="424" alt="image" src="https://github.com/realm/SwiftLint/assets/8419048/06f7a4af-62b8-4084-ad40-ea934d6470f7">

But the code left behind locally is a mess:

<img width="840" alt="image" src="https://github.com/realm/SwiftLint/assets/8419048/d808b3cc-0b1c-4c39-aa25-5c6aed88e043">

<img width="613" alt="image" src="https://github.com/realm/SwiftLint/assets/8419048/a0bcec15-f788-4658-ac13-5957461bf13a">

In other words, it took the `MyTest2` enum that I wanted to commit, formatted it nicely and committed it, but then it didn't apply the changes (`stdout`) back to my file (`test.swift`), which I believe is what's partially the responsibility of the filepath parameter (`{}`) received by other linters (e.g. prettier, SwiftFormat, etc).

My gut feeling tells me that in order for this to work, SwiftLint would have to support the filepath parameter 🤔 not sure, though.

**Comment by SimplyDanny** [comment]
Hm ... I think there must be another issue. git-format-staged warns:

> Do not attempt to read or write to `{}` in your formatter command! The placeholder exists only for referencing the file name and path.

As I've written in my previous comment, SwiftLint doesn't require the path, so we can just leave it out. Writing anything back to the path would be an error.
